scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/EcdhPskServerKeyExchange.java

```diff
@@ -0,0 +1,310 @@
+	}
+	
+	/**
+	 * Creates a new key exchange message with ServerDHParams.
+	 * 
+	 * @see <a href="http://tools.ietf.org/html/rfc4279#section-3">RFC 4279</a>
+	 * @param ecdhe - {@code ECDHECryptography}
+	 * @param clientRandom - nonce
+	 * @param serverRandom - nonce
+	 * @param namedCurveId - ec curve used 
+	 * @param peerAddress
+	 * @throws NullPointerException if any of arguments ecdhe, 
+	 * clientRandom and serverRandom are {@code null}
+	 */
+	public EcdhPskServerKeyExchange(ECDHECryptography ecdhe, Random clientRandom, Random serverRandom, 
+			int namedCurveId, InetSocketAddress peerAddress) {
+		this(null, ecdhe, clientRandom, serverRandom, namedCurveId, peerAddress);
+	}
+	
+	private EcdhPskServerKeyExchange(byte[] hintEncoded, int curveId, byte[] pointEncoded, InetSocketAddress peerAddress) throws HandshakeException {		
+		super(peerAddress);
+		this.curveId = curveId;
+		this.hintEncoded = hintEncoded;
+		this.hint = new String(this.hintEncoded, UTF_8);
+		if (pointEncoded == null) {
+			throw new NullPointerException("ephemeral public key cannot be null");
+		}
+		this.pointEncoded = Arrays.copyOf(pointEncoded, pointEncoded.length);
+		// re-create public key from params
+		SupportedGroup group = SupportedGroup.fromId(curveId);
+		if (group == null || !group.isUsable()) {
+			throw new HandshakeException(
+				String.format("Server used unsupported elliptic curve (%d) for ECDH", curveId),
+				new AlertMessage(AlertLevel.FATAL, AlertDescription.HANDSHAKE_FAILURE, peerAddress));
+		} else {
+			try {
+				point = ECDHECryptography.decodePoint(pointEncoded, group.getEcParams().getCurve());
+				KeyFactory keyFactory = KeyFactory.getInstance(KEYPAIR_GENERATOR_INSTANCE);
+				publicKey = (ECPublicKey) keyFactory.generatePublic(new ECPublicKeySpec(point, group.getEcParams()));
+			} catch (GeneralSecurityException e) {
+				LOGGER.debug("Cannot re-create server's public key from params", e);
+				throw new HandshakeException(
+					String.format("Cannot re-create server's public key from params: %s", e.getMessage()),
+					new AlertMessage(AlertLevel.FATAL, AlertDescription.INTERNAL_ERROR, peerAddress));
+			}
+		}
+	}
+
+	// TODO this is called 4 times for Flight 4
+	@Override
+	public byte[] fragmentToByteArray() {
+		DatagramWriter writer = new DatagramWriter();
+		switch (curveType) {
+		// TODO add support for other curve types
+		case EXPLICIT_PRIME:
+		case EXPLICIT_CHAR2:
+			break;
+		case NAMED_CURVE:
+			writeNamedCurve(writer);
+			break;
+		default:
+			LOGGER.warn(MSG_UNKNOWN_CURVE_TYPE, curveType);
+			break;
+		}
+		return writer.toByteArray();
+	}
+	
+	/** 
+	 * ServerKeyExchange message is always sent even if psk identity hint is not present.
+	 * 
+	 * @see <a href="https://tools.ietf.org/html/rfc5489#section-2">RFC 5489</a> 
+	 * */
+	private void writeNamedCurve(DatagramWriter writer) {
+		/*when psk hint is not present then hintEncoded length is 0 */
+		writer.write(hintEncoded.length, IDENTITY_HINT_LENGTH_BITS);
```

Comment:
writing the identity hint in this method is very confusing (and potentially wrong) because it should be done regardless of the type of curve, shouldn't it? IMHO it should be moved up into *fragmentToByteArray* or into its own method ...